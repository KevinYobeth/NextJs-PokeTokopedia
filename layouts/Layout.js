import Navbar from '../components/Navbar'

export default function Layout({ children }) {
  return (
    <>
      <Navbar />
      <main
        className='pt-24 mx-auto 
        sm:max-w-screen-sm 
        md:max-w-screen-md 
        lg:max-w-screen-lg 
        px-5'
      >{children}</main>
    </>
  );
}