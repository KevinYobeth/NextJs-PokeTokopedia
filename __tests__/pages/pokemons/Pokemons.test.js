import { render, screen } from '@testing-library/react';
import { mockCatchedPokemons, mockLocalStorageAfterRemoved, mockLocalStorageCatchedPokemons }
  from '../../../fixtures/PokemonFixtures';
import Layout from '../../../layouts/Layout';
import Pokemons from '../../../pages/pokemons';
import App, { PokemonContext } from '../../../pages/_app';

const mockPush = jest.fn();

jest.mock('next/router', () => ({
  useRouter() {
    return ({
      push: mockPush,
    });
  },
}));
jest.mock('axios');

describe('Pokemons', () => {
  let mockLocalStorage = {
    catchedPokemons: mockLocalStorageCatchedPokemons,
  };

  beforeAll(() => {
    global.Storage.prototype.setItem = jest.fn((key, value) => {
      mockLocalStorage[key] = value;
    })
    global.Storage.prototype.getItem = jest.fn((key) => mockLocalStorage[key]);
  });

  const renderComponent = (catchedPokemons = [], handleRemovePokemon = null) => render(
    <PokemonContext.Provider value={{ catchedPokemons, handleRemovePokemon }}>
      <Layout>
        <Pokemons />
      </Layout>
    </PokemonContext.Provider>
  );

  const renderComponentWithActualContext = () => {
    const props = {};
    return render(
      <App Component={Pokemons} pageProps={{ ...props }} />
    );
  };

  describe('#render', () => {
    it('should render owned pokemons page correctly when loaded', () => {
      const { container } = renderComponent(mockCatchedPokemons);

      expect(container).toMatchSnapshot();
      expect(screen.getByText('kevin')).toBeInTheDocument();
    });

    it('should render text correctly when loaded with no catched pokemons', () => {
      renderComponent();
      const expectedText = 'You don\'t have any pokemon yet';

      expect(screen.getByText(expectedText)).toBeInTheDocument();
    });
  });

  describe('#onClick', () => {
    it('should redirect to pokemon details page when details button is clicked', () => {
      renderComponent(mockCatchedPokemons);
      const detailsButton = screen.getAllByText('Details')[0];

      detailsButton.click();

      expect(mockPush).toHaveBeenCalledWith('/pokemons/pikachu');
    });

    it('should release pokemon when release button is clicked', () => {
      const mockHandleRemovePokemon = jest.fn();
      renderComponent(mockCatchedPokemons, mockHandleRemovePokemon);
      const releaseButton = screen.getAllByText('Release')[1];

      releaseButton.click();

      expect(mockHandleRemovePokemon).toHaveBeenCalledWith('second-kevin');
    });

    it('should call handleRemovePokemon from app.js when pokemon is released', async () => {
      renderComponentWithActualContext();
      const releaseButton = screen.getAllByText('Release')[0];

      releaseButton.click();

      expect(localStorage.setItem)
        .toHaveBeenCalledWith('catchedPokemons', mockLocalStorageAfterRemoved);
    });
  });
});