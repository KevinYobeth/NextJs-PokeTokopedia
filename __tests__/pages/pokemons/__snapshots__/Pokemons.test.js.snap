// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemons #render should render owned pokemons page correctly when loaded 1`] = `
<div>
  <div
    class="fixed flex items-center justify-center w-full h-20 bg-white drop-shadow-lg 
 z-30"
  >
    <div
      class="flex items-center justify-between mx-auto w-full p-5 sm:max-w-screen-sm
 md:max-w-screen-md lg:max-w-screen-lg"
    >
      <a
        href="/"
      >
        <h1
          class="text-4xl font-bold"
        >
          POKE
        </h1>
      </a>
      <a
        data-testid="pokeball"
        href="/pokemons"
      >
        <div
          class="flex items-center gap-x-2"
        >
          <p>
            (
            3
            )
          </p>
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDAiIGhlaWdodD0iNDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="my pokemon"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
      </a>
    </div>
  </div>
  <main
    class="pt-24 mx-auto 
 sm:max-w-screen-sm 
 md:max-w-screen-md 
 lg:max-w-screen-lg 
 px-5"
  >
    <div
      class="grid grid-cols-2 sm:grid-cols-4 md:grid-cols-6 gap-3 mb-10"
    >
      <div
        class="col-span-2 w-full bg-white rounded-md drop-shadow-md 
 p-5 gap-y-3"
      >
        <div
          class="flex items-center justify-center -mt-4"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTEyIiBoZWlnaHQ9IjUxMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="pikachu"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="flex flex-col justify-center col-span-2 gap-y-1"
        >
          <div
            class="font-poppins font-medium text-center"
          >
            <h2
              class="text-xl"
            >
              Pikachu
              #
              25
            </h2>
            <p>
              pika
            </p>
            <p
              class="font-poppins font-normal text-sm"
              data-testid="owned-count"
            >
              Owned: 
              <span>
                1
              </span>
            </p>
          </div>
        </div>
        <div
          class="flex flex-col gap-y-2 col-span-2 mt-5 order-3"
        >
          <button
            class="w-full rounded-md p-3 bg-red-500 hover:bg-red-600 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
          >
            <p
              class="font-poppins text-center"
            >
              Release
            </p>
          </button>
          <button
            class="w-full rounded-md p-3 bg-gray-800 hover:bg-gray-900 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
          >
            <p
              class="font-poppins text-center"
            >
              Details
            </p>
          </button>
        </div>
      </div>
      <div
        class="col-span-2 w-full bg-white rounded-md drop-shadow-md 
 p-5 gap-y-3"
      >
        <div
          class="flex items-center justify-center -mt-4"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTEyIiBoZWlnaHQ9IjUxMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="charmander"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="flex flex-col justify-center col-span-2 gap-y-1"
        >
          <div
            class="font-poppins font-medium text-center"
          >
            <h2
              class="text-xl"
            >
              Charmander
              #
              4
            </h2>
            <p>
              second-kevin
            </p>
            <p
              class="font-poppins font-normal text-sm"
              data-testid="owned-count"
            >
              Owned: 
              <span>
                2
              </span>
            </p>
          </div>
        </div>
        <div
          class="flex flex-col gap-y-2 col-span-2 mt-5 order-3"
        >
          <button
            class="w-full rounded-md p-3 bg-red-500 hover:bg-red-600 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
          >
            <p
              class="font-poppins text-center"
            >
              Release
            </p>
          </button>
          <button
            class="w-full rounded-md p-3 bg-gray-800 hover:bg-gray-900 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
          >
            <p
              class="font-poppins text-center"
            >
              Details
            </p>
          </button>
        </div>
      </div>
      <div
        class="col-span-2 w-full bg-white rounded-md drop-shadow-md 
 p-5 gap-y-3"
      >
        <div
          class="flex items-center justify-center -mt-4"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTEyIiBoZWlnaHQ9IjUxMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="charmander"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="flex flex-col justify-center col-span-2 gap-y-1"
        >
          <div
            class="font-poppins font-medium text-center"
          >
            <h2
              class="text-xl"
            >
              Charmander
              #
              4
            </h2>
            <p>
              kevin
            </p>
            <p
              class="font-poppins font-normal text-sm"
              data-testid="owned-count"
            >
              Owned: 
              <span>
                2
              </span>
            </p>
          </div>
        </div>
        <div
          class="flex flex-col gap-y-2 col-span-2 mt-5 order-3"
        >
          <button
            class="w-full rounded-md p-3 bg-red-500 hover:bg-red-600 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
          >
            <p
              class="font-poppins text-center"
            >
              Release
            </p>
          </button>
          <button
            class="w-full rounded-md p-3 bg-gray-800 hover:bg-gray-900 text-white -translate-y-1.5 
      hover:-translate-y-2.5 active:-translate-y-1 duration-300 active:duration-75 
      transition-transform"
          >
            <p
              class="font-poppins text-center"
            >
              Details
            </p>
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;
